onboarding/mlsys/1-backprop.md

```diff
@@ -0,0 +1,414 @@
+print("mfinal", mfinal)
+```
+
+Explain in words what is happening.
+
+## Batch Size in Training
+
+Your training fuction calculated the gradient for each data point, taking a
+step to minimize the loss for that particular point, then moving on to the next
+point.  This strategy (considering exactly one batch at a time) is called
+"stochastic gradient descent".  It can be an messy process because real gradient
+curves are not smooth parabolas like in this example, and taking a step every
+data point can cause you to take a lot of steps in the overall wrong direction.
+
+There is also "batch gradient descent", which is where you calculate the total
+gradient for your whole dataset (adding up the gradients for every data point
+in the whole dataset) before taking a single step.  This is a more reliable way
+to minimize the loss than stoachastic gradient descent because by looking at
+all data points before stepping, you can be sure you're stepping in the overall
+right direction.  However, nobody in deep learning really does this, because the
+datasets are normally huge and it would take forever to train a model on a
+dataset if you could only step once per epoch.
+
+There is an in-between option, which is more accurate than stochastic gradient
+descent but much fatser than batch gradient descent.  It is called "mini-batch
+gradient descent", and it's what basically the only gradient descent used in deep
+learning.  In mini-batch gradient descent, you have have some `batch_size`,
+usually a multiple of 2.  In each iteration of training, you select a
+mini-batch of `batch_size` elements, calculate the gradients for that
+mini-batch, take a step, and then repeat with the next mini-batch.  Mini-batch
+gradient descent is is so universal in our industry that you won't hear anybody
+talk about it, and the word "batch" is widely preferred to "mini-batch".
+
+Mini-batch training is especially well-suited to training on GPUs.  Graphics
+cards are parallel-processing engines, and since the total gradient for a
+batch (no use calling it a "mini-batch" around here) can be calculated in
+paralell (by independently calculating the gradients from each data point, and
+summing or averaging them together at the end), the time to do a forward and
+backward pass on a single data point is approximately equal to the time to do a
+forward and backward pass on a whole batch of data.  However, since you are
+considering multiple data points for each step, you get a more accurate step in
+the same amount of training time.  The upper limit of batch size is how much
+GPU memory you have.  Larger batch sizes means you are doing more parallel
+calculations at the same time, which requires more memory.
+
+### Q8
+
+Write a variation of your previous training loop, but this time do mini-batch
+training and add a `batch_size` parameter.  Note that when you combine
+gradients, it is most common to average them, not just to sum them.
```

Comment:
When you combine gradients, it is more common to average them than to just sum them.